### ISSUE TITLE
Model.evaluate retuns 0 loss when using cutom model class

### ISSUE THREAD

**Issue by arielAmsel** [keras-team/keras#14705]
I am trying to use my own train step in with Keras by creating a class that inherits from Model. It seems that the training works correctly but the evaluate function always returns 0 on the loss even if I send to it the train data, which have a big loss value during the training.
I can't share my code but was able to reproduce using the example form the Keras api in <https://keras.io/guides/customizing_what_happens_in_fit/> I changed the Dense layer to have 2 units instead of one, made its activation to sigmoid and the loss and metrics are compiled.

The code:
```
import tensorflow.keras as keras
import tensorflow as tf
import numpy as np

loss_tracker = keras.metrics.Mean(name="loss")
# mae_metric = keras.metrics.MeanAbsoluteError(name="mae")

class CustomModel(keras.Model):
    def train_step(self, data):
        x, y = data

        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)  # Forward pass
            # Compute our own loss
            loss = self.compiled_loss(y, y_pred)

        # Compute gradients
        trainable_vars = self.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)

        # Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))

        # Compute our own metrics
        loss_tracker.update_state(loss)
        self.compiled_metrics.update_state(y, y_pred)
        return {"loss": loss_tracker.result(), "mae": self.compiled_metrics.metrics[0].result()}

    @property
    def metrics(self):
        # We list our `Metric` objects here so that `reset_states()` can be
        # called automatically at the start of each epoch
        # or at the start of `evaluate()`.
        # If you don't implement this property, you have to call
        # `reset_states()` yourself at the time of your choosing.
        return [loss_tracker] + self.compiled_metrics.metrics


if __name__ == '__main__':
    # Construct an instance of CustomModel
    inputs = keras.Input(shape=(32,))
    outputs = keras.layers.Dense(2, activation='sigmoid')(inputs)
    model = CustomModel(inputs, outputs)

    # We don't passs a loss or metrics here.
    model.compile(optimizer="adam", loss='mean_squared_error', metrics=['mean_squared_error'])

    # Just use `fit` as usual -- you can use callbacks, etc.
    x = np.random.random((1000, 32))
    y = np.random.random((1000, 1))
    model.fit(x, y, epochs=5)
    print(model.evaluate(x, y, return_dict=True))
    x = np.random.random((1000, 32))
    y = np.random.random((1000, 1))
    print(model.evaluate(x, y, return_dict=True))

```

I try to run evaluate on the original train data as well on some random data and both return 0 on the loss and MAE.

The output:
Epoch 1/5<br>
32/32 [==============================] - 0s 694us/step - loss: 0.1016 - mae: 0.2704<br>
Epoch 2/5<br>
32/32 [==============================] - 0s 678us/step - loss: 0.0965 - mae: 0.2666<br>
Epoch 3/5<br>
32/32 [==============================] - 0s 650us/step - loss: 0.0960 - mae: 0.2661<br>
Epoch 4/5<br>
32/32 [==============================] - 0s 676us/step - loss: 0.0956 - mae: 0.2659<br>
Epoch 5/5<br>
32/32 [==============================] - 0s 739us/step - loss: 0.0951 - mae: 0.2652<br>
Evaluations:
32/32 [==============================] - 0s 914us/step - loss: 0.0000e+00 - mae: 0.0000e+00<br>
{'loss': 0.0, 'mae': 0.0}<br>
32/32 [==============================] - 0s 936us/step - loss: 0.0000e+00 - mae: 0.0000e+00<br>
{'loss': 0.0, 'mae': 0.0}<br>

The first evaluation is on the training data and returns 0 although in the training it isn't 0 and also the random data returns 0 - no way it is really 0.

Any ideas what might cause this problem? Or how to solve this except for overriding myself the eval method?
